### DISCUSSION TITLE
Editor Inheritence Nesting

### DISCUSSION THREAD

**Discussion by Bodyclock-Games** [stride3d/stride#1931] [Q&A] [answered]
Hi All, sorry for the incessant questions.
I have a Terrain Manager startup editor sript that sets the noise parameters for procedural terrain. The are 4 levels of abstract nesting (see below). All works fine in the editor. But at runtime, I can only access levels 1 & 2, level 3 is always null even when its set in the editor. I can set it in code no problem.
Is there a restriction on the levels of serialization or on the levels of reflectance on editor scripts? Is there something I'm missing?
Cheers
Chris

![TerrainDialog](https://github.com/stride3d/stride/assets/49901560/5cfa22d8-faec-4611-90ac-28dee07a952b)

**Comment by Eideren** [comment]
Can you send the `Standard` and `Perlin` class definition as well as the `Fractal` member definition ? Hard to tell without source.
Also, you might want to look at the new and not yet released page of the wiki on serialization over here https://github.com/stride3d/stride-docs/blob/master/en/manual/scripts/serialization.md

**Comment by Bodyclock-Games** [comment]
Thanks for getting back to me. While writing this I did wonder if it has to do with setting the default values. But I use the same method in Level 1 and 2. 


The classes all follow the same pattern. All data except for a method to set the noise parameters.
Level 1 is the TerrainType, in this case Procedural

```
using Bodyclock.Stride.Core.Mathematics;
using Stride.Core;
using Stride.Engine;

namespace Bodyclock.Stride.Terrain
{
    [DataContract]
    public class Procedural : TerrainType
    {
        public NoiseType NoiseType;

        public override ModelComponent Initialise(TerrainManager terrainManager)
        `{`
            NoiseType.SetNoise(terrainManager.Noise);
            return modelComponent;
        }
    }
}
```
Level 2 is the  NoiseType abstract class and Perlin etc inherit from that.

```
using Stride.Core;
using Stride.Core.Annotations;

namespace Bodyclock.Stride.Core.Mathematics
{
    public abstract class NoiseType
    {
        [DataMember(140)]
        [Display("Fractal")]
        public FractalType FractalType;

        [DataMember(200)]
        [Display("Domain Warp")]
        public DomainWarpType DomainWarpType;

        public abstract Noise.NoiseType Type { get; }

        [DataMember(100)]
        [Display("Seed")]
        public uint Seed { get; set; } = 123456789;

        [DataMember(110)]
        [DataMemberRange(0, 1, 0.001, 0.1, 3)]
        [Display("Frequency")]
        public float Frequency { get; set; } = 0.01f;

        public virtual void SetNoise(Noise noise)
        {
            noise.SetNoiseType(Type);
            noise.SetSeed((int)Seed);
            noise.SetFrequency(Frequency);
            if (FractalType is not null) FractalType.SetNoise(noise);
            if (DomainWarpType is not null) DomainWarpType.SetNoise(noise);
        }
    }
}
```

Concrete noise type.

```
using Stride.Core;

namespace Bodyclock.Stride.Core.Mathematics
{
    [DataContract]
    public class Perlin : NoiseType
    {
        public override Noise.NoiseType Type { get; } = Noise.NoiseType.Perlin;
    }
}
```

Level 3 is the FractalType and the DomainWarpType abstract classes. The are the same structure.
```

using Stride.Core;
using Stride.Core.Annotations;

namespace Bodyclock.Stride.Core.Mathematics
{
    public abstract class FractalType
    {
        public abstract Noise.FractalType Type { get; }

        [DataMember(100)]
        [DataMemberRange(1, 10, 1, 1, 0)]
        [Display("Octaves")]
        public uint Octaves { get; set; } = 2;

        [DataMember(110)]
        [DataMemberRange(0, 10, 0.001, 0.1, 3)]
        [Display("Lacunarity")]
        public float Lacunarity { get; set; } = 2.0f;

        [DataMember(120)]
        [DataMemberRange(0, 10, 0.001, 0.1, 3)]
        [Display("Gain")]
        public float Gain { get; set; } = 0.5f;

        [DataMember(130)]
        [DataMemberRange(0, 1, 0.001, 0.1, 3)]
        [Display("Weight")]
        public float Weight { get; set; } = 0.5f;

        public virtual void SetNoise(Noise noise)
        {
            noise.SetFractalType(Type);
            noise.SetFractalOctaves((int)Octaves);
            noise.SetFractalLacunarity(Lacunarity);
            noise.SetFractalGain(Gain);
            noise.SetFractalWeightedStrength(Weight);
        }
    }
}
```

And the concrete class

```
using Stride.Core;

namespace Bodyclock.Stride.Core.Mathematics
{
    [DataContract]
    [Display("Standard")]
    public class BrownianMotion : FractalType
    {
        public override Noise.FractalType Type { get; } = Noise.FractalType.FBm;
    }
}
```

There is a Level 4 that just follows the same pattern..

**Reply by manio143** [reply, marked as answer]
You need to put DataContract on your abstract classes as well. Otherwise we don't generate  serializers for those classes and any member defined in your base class is omitted from runtime.

**Reply by Bodyclock-Games** [reply]
Thank you. That was it. i appreciate the patience of the forums with someone new to Stride.  :)
